src/components/calcite-action-menu/calcite-action-menu.tsx

```diff
@@ -0,0 +1,338 @@
+      action.id = id;
+    }
+
+    action.active = index === activeMenuItemIndex;
+  };
+
+  updateActions = (actions: HTMLCalciteActionElement[]): void => {
+    actions?.forEach(this.updateAction);
+  };
+
+  getActions = (): void => {
+    const actions =
+      (this.el
+        .querySelector("slot")
+        ?.assignedElements({ flatten: true })
+        .filter((el) => el.tagName === "CALCITE-ACTION") as HTMLCalciteActionElement[]) || [];
+
+    this.updateActions(actions);
+
+    this.actions = actions;
+  };
+
+  isValidKey(key: string, supportedKeys: string[]): boolean {
+    return !!supportedKeys.find((k) => k === key);
+  }
+
+  menuButtonKeyUp = (event: KeyboardEvent): void => {
+    const { key } = event;
+    const { actions } = this;
+
+    if (!this.isValidKey(key, SUPPORTED_BUTTON_NAV_KEYS)) {
+      return;
+    }
+
+    event.preventDefault();
+
+    if (!actions.length) {
+      return;
+    }
+
+    this.toggleOpen(true);
+    this.handleActionNavigation(key, actions);
+  };
+
+  menuButtonKeyDown = (event: KeyboardEvent): void => {
+    const { key } = event;
+
+    if (!this.isValidKey(key, SUPPORTED_BUTTON_NAV_KEYS)) {
+      return;
+    }
+
+    event.preventDefault();
+  };
+
+  menuActionsContainerKeyDown = (event: KeyboardEvent): void => {
+    const { key } = event;
+    const { actions, activeMenuItemIndex } = this;
+
+    if (key === "Tab") {
+      this.open = false;
+      return;
+    }
+
+    if (key === " " || key === "Enter") {
+      event.preventDefault();
+      const action = actions[activeMenuItemIndex];
+      action ? action.click() : this.toggleOpen(false);
+      return;
+    }
+
+    if (this.isValidKey(key, SUPPORTED_MENU_NAV_KEYS)) {
+      event.preventDefault();
+    }
+  };
+
+  menuActionsContainerKeyUp = (event: KeyboardEvent): void => {
+    const { key } = event;
+    const { actions } = this;
+
+    if (key === "Escape") {
+      this.toggleOpen(false);
+      return;
+    }
+
+    if (!this.isValidKey(key, SUPPORTED_MENU_NAV_KEYS)) {
+      return;
+    }
+
+    event.preventDefault();
+
+    if (!actions.length) {
+      return;
+    }
+
+    this.handleActionNavigation(key, actions);
+  };
+
+  handleActionNavigation = (key: string, actions: HTMLCalciteActionElement[]): void => {
+    const currentIndex = this.activeMenuItemIndex;
+
+    if (key === "Home") {
+      this.activeMenuItemIndex = 0;
+    }
+
+    if (key === "End") {
+      this.activeMenuItemIndex = actions.length - 1;
+    }
+
+    if (key === "ArrowUp") {
+      this.activeMenuItemIndex = getRoundRobinIndex(Math.max(currentIndex - 1, -1), actions.length);
+    }
+
+    if (key === "ArrowDown") {
+      this.activeMenuItemIndex = getRoundRobinIndex(currentIndex + 1, actions.length);
+    }
+  };
+
+  toggleOpen = (value = !this.open): void => {
+    this.open = value;
+    this.activeMenuItemIndex = -1;
+
+    if (value) {
+      setTimeout(() => focusElement(this.menuEl), MENU_ANIMATION_DELAY_MS);
```

Comment:
Not crucial for this PR, but could this use the `animationend`/`transitionend` event to avoid specifying and maintaining the delay?

Reply:
That would be awesome. Is there a way I can do it though? The animation is internal to the popover component. I would have to go through shadow dom into it's internals right?

Reply:
Good question. I would have to test it out. I'm not sure those events interact with the shadow DOM boundary.

Reply:
We might have to expose some kind of event so other components could listen? I know i've seen other cases where we are using setTimeout to handle focusing after an animation.

- https://github.com/Esri/calcite-components/blob/cd59fdac26135cb5906779568d24ed6812e1a994/src/components/calcite-modal/calcite-modal.tsx#L297
- https://github.com/Esri/calcite-components/blob/cd59fdac26135cb5906779568d24ed6812e1a994/src/components/calcite-inline-editable/calcite-inline-editable.tsx#L234
- etc: https://github.com/Esri/calcite-components/search?l=TSX&q=setTimeout%28

Reply:
Yeah, sounds like a new pattern is a-brewin. 😄  Let's tackle this in a new issue then since it looks like we need to do some prototype work. I did a quick test and you are right in that we'd need to process the event at the shadow DOM level and then surface it.

Reply:
We can create some naming standard for private animation events 👍

Reply:
Created issue.